### PR TITLE
solc: 0.5.7 -> 0.5.8

### PR DESCRIPTION
**Pull request by sifmelcara**
###### Things done

1. Disable shared library installation. I don't really see use cases that need expose `solc`'s internal components... they are not meant to be used by compiler users.
2. Previously I only added basic unit tests to `checkPhase`, this time I enables all tests except tests that need IPC or SMT. `ncurses` and `python2` added as dependency for the test scripts.

cc maintainers @dbrock @akru @lionello

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Lassulus** [comment]
@GrahamcOfBorg build solc

**Comment by akru** [comment]
@sifmelcara Could you keep shared libraries, please. May be in .lib location. It’s required for http://hackage.haskell.org/package/web3 package.

**Comment by sifmelcara** [comment]
Sorry, could you elaborate a little bit more? I can install web3 without installing solc just fine (using `stack --nix --nix-packages zlib install web3`). Also from my understanding, web3 should have nothing to do with `solc`?

Thanks!

**Comment by akru** [comment]
As you can see, hs-web3 has special flag for enabling experimental FFI bindings to libsolidity.

**Comment by sifmelcara** [comment]
Ah, ok. So I made shared library build optional and tested it using `nix-build -E "(import ./default.nix {}).solc.override { buildSharedLibrary = true; }"`. It works fine for me.

Now I also enabled tests on darwin, but I have no MacOS machine to try if the tests works correctly on it. @Lassulus could you ask @GrahamcOfBorg to built it again? Thank you so much.

**Comment by Lassulus** [comment]
@GrahamcOfBorg build solc

**Comment by akru** [comment]
@sifmelcara Overriding isn’t easy to use in practice outside of nix files, in stack for example. What’s about solc.lib destination for shared libraries?

**Comment by sifmelcara** [comment]
Hmm... IMHO we should not build `solc`'s internal components as shared library, as this is not a supported way to use the compiler.

But if you want, we can keep building shared library by default for now. Then we can make `web3` use `solc`'s standard JSON interface instead of using C FFI. This solution can also make the solidity compilation feature in `web3` available in other distros. What do you think? (I can try to implement this if you want)

**Comment by akru** [comment]
In my personal opinion linking is better than process spawning. Ok, if you want - do it. I’ll rewrite compiler module in the future.

**Comment by sifmelcara** [comment]
Force pushed. I kept the configurations besides adding tests, so this should work for `web3-hs`.

Yes of course linking is better than process spawning. I just feels a little bit annoying that it require non-nix users to apply custom patch (`shared-libs-install.patch`), use custom cmake flag and rebuild `solc` themself. Anyway, thank you for answering my questions and all your efforts  :)

**Comment by akru** [comment]
Basically I focus on stack build with nix integration for experimental features.

Thank you for your work @sifmelcara.

**Comment by akru** [comment]
@GrahamcOfBorg build solc

**Comment by lionello** [comment]
Fails on Darwin:
```
[ 11%] Linking CXX shared library liblangutil.dylib
Undefined symbols for architecture x86_64:
  "dev::Exception::what() const", referenced from:
      vtable for langutil::InternalCompilerError in CharStream.cpp.o
      vtable for boost::exception_detail::clone_impl<langutil::InternalCompilerError> in CharStream.cpp.o
      vtable for langutil::FatalError in ErrorReporter.cpp.o
      vtable for langutil::Error in ErrorReporter.cpp.o
      vtable for boost::exception_detail::clone_impl<langutil::FatalError> in ErrorReporter.cpp.o
      vtable for langutil::Error in Exceptions.cpp.o
      vtable for langutil::InternalCompilerError in ParserBase.cpp.o
      ...
  "typeinfo for dev::Exception", referenced from:
      typeinfo for langutil::InternalCompilerError in CharStream.cpp.o
      typeinfo for langutil::FatalError in ErrorReporter.cpp.o
      typeinfo for langutil::Error in ErrorReporter.cpp.o
      typeinfo for langutil::Error in Exceptions.cpp.o
      typeinfo for langutil::InternalCompilerError in ParserBase.cpp.o
      typeinfo for langutil::InternalCompilerError in Scanner.cpp.o
      typeinfo for langutil::SemVerError in SemVerHandler.cpp.o
      ...
ld: symbol(s) not found for architecture x86_64
clang-5.0: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [liblangutil/CMakeFiles/langutil.dir/build.make:224: liblangutil/liblangutil.dylib] Error 1
make[1]: *** [CMakeFiles/Makefile2:222: liblangutil/CMakeFiles/langutil.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
builder for '/nix/store/hn7dilxz4ryrc22vbi7hfz12jxncy10h-solc-0.5.8.drv' failed with exit code 2
error: build of '/nix/store/hn7dilxz4ryrc22vbi7hfz12jxncy10h-solc-0.5.8.drv' failed
```

**Comment by sifmelcara** [comment]
Hmm.. that's strange. This PR only adds tests on linux platform and updates `solc` version. Also Grahamcofborg builds fine on darwin before.

Maybe the first thing we should do is to ask @GrahamcOfBorg to perform build on Darwin to confirm the issue? As I don't have MacOS to test it.

**Comment by lionello** [comment]
@sifmelcara My guess is we need to use `boost17x`. Testing now.

**Comment by lionello** [comment]
No, that didn't work

File: pkgs/development/compilers/solc/default.nix [hunk 1]
```diff
@@ -42,13 +40,33 @@ stdenv.mkDerivation {
...
-  buildInputs = [ boost ]
-    ++ stdenv.lib.optionals z3Support [ z3 ];
+  buildInputs = [ boost ] ++ stdenv.lib.optionals z3Support [ z3 ];
+  checkInputs = [ ncurses python2 ];
+
+  # Test fails on darwin for unclear reason
+  doCheck = stdenv.hostPlatform.isLinux;
+
+  checkPhase = ''
+    for dir in $(find . -type d);
```
Comment (asymmetric) [inline pkgs/development/compilers/solc/default.nix]: I don't think you need the `;` here, unless you put `do` on the same line.

Comment (asymmetric) [reply pkgs/development/compilers/solc/default.nix]: Also this raises the [shellcheck SC2044 warning](https://github.com/koalaman/shellcheck/wiki/SC2044).

File: pkgs/development/compilers/solc/default.nix [hunk 2]
```diff
@@ -42,13 +40,33 @@ stdenv.mkDerivation {
...
+    do
+      export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$(pwd)/$dir
```
Comment (asymmetric) [inline pkgs/development/compilers/solc/default.nix]: SC2155

File: pkgs/development/compilers/solc/default.nix [hunk 3]
```diff
@@ -42,13 +40,33 @@ stdenv.mkDerivation {
...
+    done
+    pushd ..
+    # IPC tests need aleth avaliable, so we disable it
+    sed -i "s/IPC_ENABLED=true/IPC_ENABLED=false\nIPC_FLAGS=\"--no-ipc\"/" ./scripts/tests.sh
+    for i in ./scripts/*.sh; do
+      patchShebangs $i
```
Comment (asymmetric) [inline pkgs/development/compilers/solc/default.nix]: SC2086

File: pkgs/development/compilers/solc/default.nix [hunk 4]
```diff
@@ -42,13 +40,33 @@ stdenv.mkDerivation {
...
+    done
+    for i in ./scripts/*.py; do
+      patchShebangs $i
```
Comment (asymmetric) [inline pkgs/development/compilers/solc/default.nix]: SC2086

File: pkgs/development/compilers/solc/default.nix [hunk 5]
```diff
@@ -42,13 +40,33 @@ stdenv.mkDerivation {
...
+    done
+    for i in ./test/*.sh; do
+      patchShebangs $i
```
Comment (asymmetric) [inline pkgs/development/compilers/solc/default.nix]: SC2086

File: pkgs/development/compilers/solc/default.nix [hunk 6]
```diff
@@ -42,13 +40,33 @@ stdenv.mkDerivation {
...
+    done
+    TERM=xterm ./scripts/tests.sh
+    popd
```
Comment (asymmetric) [inline pkgs/development/compilers/solc/default.nix]: Is this necessary?

Either way, this raises SC2164, although not sure how important it is for us.

**Comment by asymmetric** [comment]
We could [confirm](https://travis-ci.com/dapphub/dapptools/jobs/199001100) the issue with 0.5.8 on Darwin on our CI server (building [this PR](https://github.com/NixOS/nixpkgs/pull/61206))

**Comment by sifmelcara** [comment]
I just look through solidity's `git diff v0.5.7 v0.5.8`, and looks like there is no obvious thing that will cause link failure. I guess someone have access to MacOS need to do `git bisect` to find out the problematic commit, or we can simply not patch our custom shared library patch when build on MacOS.

File: pkgs/development/compilers/solc/default.nix (hunk 1 above)
Comment (sifmelcara) [reply pkgs/development/compilers/solc/default.nix]: Ok will fix these warnings

File: pkgs/development/compilers/solc/default.nix (hunk 6 above)
Comment (sifmelcara) [reply pkgs/development/compilers/solc/default.nix]: We always `pushd ..`, so it is probably fine

**Comment by sifmelcara** [comment]
@akru are you ok with not applying shared library install patch on MacOS?

**Comment by akru** [comment]
@sifmelcara yes, when supporting it makes a trouble we can just disable it for Darwin.

**Comment by sifmelcara** [comment]
Rebased and disabled shared library build on Darwin. @Lassulus could you trigger @GrahamcOfBorg 's darwin build to confirm everything is ok? Thank you!

>  I think it’s not so bad idea to move shared libs to separate package, for example with ‘libsolidity’ name. We can set Linux only target for it.

Maybe we should ask upstream to provide official support of shared library build (to provide a way to use the compiler as a library). I can ping you when I open an upstream issue.

**Comment by akru** [comment]
@sifmelcara Ok

@GrahamcOfBorg build solc

**Comment by sifmelcara** [comment]
@akru IIRC you need to be trusted by @GrahamcOfBorg to build on MacOS. Can someone do that to confirm the build is ok on MacOS?

**Comment by Lassulus** [comment]
@GrahamcOfBorg build solc

**Comment by sifmelcara** [comment]
Everything looks fine, let's merge?

**Pull request merged**